packages/pricing/src/services/pricing-module.ts

```diff
@@ -247,19 +266,377 @@ export default class PricingModuleService<
       count,
     ]
   }
+  @InjectManager("baseRepository_")
+  async create(
+    data: PricingTypes.CreatePriceSetDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<PricingTypes.PriceSetDTO[]> {
+    const priceSets = await this.create_(data, sharedContext)
+
+    return this.list(
+      { id: priceSets.filter((p) => !!p).map((p) => p!.id) },
+      {
+        relations: ["rule_types", "money_amounts", "price_rules"],
+      }
+    )
+  }
 
   @InjectTransactionManager(shouldForceTransaction, "baseRepository_")
-  async create(
+  protected async create_(
     data: PricingTypes.CreatePriceSetDTO[],
     @MedusaContext() sharedContext: Context = {}
   ) {
-    const priceSets = await this.priceSetService_.create(data, sharedContext)
+    const ruleAttributes = data
+      .map((d) => d.rules?.map((r) => r.rule_attribute) ?? [])
+      .flat()
 
-    return this.baseRepository_.serialize<PricingTypes.PriceSetDTO[]>(
-      priceSets,
+    const ruleTypes = await this.ruleTypeService_.list(
       {
-        populate: true,
+        rule_attribute: ruleAttributes,
+      },
+      {},
+      sharedContext
+    )
+
+    const ruleTypeMap = ruleTypes.reduce((acc, curr) => {
+      acc.set(curr.rule_attribute, curr)
+      return acc
+    }, new Map())
+
+    const invalidRuleAttributes = ruleAttributes.filter(
+      (r) => !ruleTypeMap.has(r)
+    )
+
+    if (invalidRuleAttributes.length > 0) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Rule types don't exist for: ${invalidRuleAttributes.join(", ")}`
+      )
+    }
+
+    const invalidMoneyAmountRule = data
+      .map(
+        (d) => d.money_amounts?.map((ma) => Object.keys(ma.rules)).flat() ?? []
+      )
+      .flat()
+      .filter((r) => !ruleTypeMap.has(r))
+
+    if (invalidMoneyAmountRule.length > 0) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Rule types don't exist for money amounts with rule attribute: ${invalidMoneyAmountRule.join(
+          ", "
+        )}`
+      )
+    }
+
+    const priceSets = await Promise.all(
+      data.map(async (d) => {
```

Comment:
Hey! I think it's a valid comment. I've been wracking my brain on how to actually achieve this since all the calls are dependent on the previous call in order to relate the entities and I can't really find a good solution. A solution could be to "manually" assign the ids for each entity as the data for each is scaffolded (manually in this case would be dependent on the beforeCreate assigning an id or extending entities with a `getId` method or something similar, don't have a neat solution for this yet). This would allow us to reference the ids of entities further down. 

Otherwise we have to depend on the ordering of what we put in to methods being the same as what is returned which seems pretty janky.

An initial take on why this approach might be okay though is that `create` is not going to the database but creating references to the entities and the database is only touched on `flush`. I know this might not be the case for other ORMs, meaning this might not be very performance if you are creating your own DAL.

would love to hear your thoughts

Reply:
yes I see what you mean, indeed having an id from the start would be the easiest way to do it. for that we could have a service method that only create the entities without persisting. Then adding @OnInit on the before create would take care of assigning the id automatically when manager.create is called. Maybe this is something we can think of in another iteration unless you think it would be quick to do. Let me know what you think about it.

ps: this might be a too early optimisation but just want your takes on it for the future if we fall into it